### PR TITLE
[stable9] lenz1111's fantastic share download range requests support

### PR DESCRIPTION
**Pull request by PVince81**
Backport of https://github.com/owncloud/core/pull/24750 to stable9

Please review and test @lenz1111 @michaelstingl @davitol @owncloud/sharing @owncloud/filesystem 

Had to revert the PSR-4 namespace of the new exception.

### CODE REVIEW

File: lib/public/files/unseekableexception.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * Files/UnseekableException class
+ */
+
+// use OCP namespace for all classes that are considered public.
+// This means that they should be used by apps instead of the internal ownCloud classes
+namespace OCP\Files;
+
+/**
+ * Exception for seek problem
+ * @since 9.1.0
```
Comment (lenz1111) [inline lib/public/files/unseekableexception.php]: Should be here @since 9.1.0?

**Comment by lenz1111** [comment]
WFM :+1:

**Comment by davitol** [comment]
WFM too :+1:

File: lib/public/files/unseekableexception.php (hunk 1 above)
Comment (PVince81) [inline lib/public/files/unseekableexception.php]: good point... maybe we need to use another approach that doesn't require an exception :frowning:

**Comment by PVince81** [comment]
:-1: we need to find a solution for the newly added exception.
Since it's added to the public namespace it qualifies as "API change" or "API addition".

@lenz1111 is there a way to make the code work without the exception ?

**Comment by lenz1111** [comment]
@PVince81 We could return false by `readFilePart` when` fseek` fail - but easiest workaround is to change thrown exception (by readFilePart) to for example `NotPermittedException` - what you think?

**Comment by PVince81** [comment]
Let's do the `NotPermittedException` one, because `return false` is for when fopen fails, in which case it doesn't make sense to retry.

**Comment by lenz1111** [comment]
OK - I'll fix it

**Comment by PVince81** [comment]
@lenz1111 protip: after committing stuff, please ping people for review because we don't get a notification from Github automatically :smile:

**Comment by PVince81** [comment]
Retested, works. :+1:

**Comment by lenz1111** [comment]
@PVince81 ok

**Pull request merged**